cluster/meta_executor.go

```diff
@@ -0,0 +1,146 @@
+package cluster
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/influxdata/influxdb"
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/influxdata/influxdb/services/meta"
+)
+
+// MetaExecutor executes meta queries on all data nodes.
+type MetaExecutor struct {
+	mu             sync.RWMutex
+	timeout        time.Duration
+	pool           *clientPool
+	maxConnections int
+	Logger         *log.Logger
+	Node           *influxdb.Node
+
+	MetaClient interface {
+		DataNode(id uint64) (ni *meta.NodeInfo, err error)
+		DataNodes() ([]meta.NodeInfo, error)
+	}
+}
+
+// NewMetaExecutor returns a new initialized *MetaExecutor.
+func NewMetaExecutor() *MetaExecutor {
+	return &MetaExecutor{
+		timeout:        DefaultWriteTimeout,
+		pool:           newClientPool(),
+		maxConnections: 1000,
+		Logger:         log.New(os.Stderr, "[meta-executor] ", log.LstdFlags),
+	}
+}
+
+// ExecuteStatement executes a single InfluxQL statement on all nodes in the cluster concurrently.
+func (m *MetaExecutor) ExecuteStatement(stmt influxql.Statement, database string) error {
+	// Get a list of all nodes the query needs to be executed on.
+	nodes, err := m.MetaClient.DataNodes()
+	if err != nil {
+		return err
+	}
+
+	// Start a goroutine to execute the statement on each of the remote nodes.
+	var wg sync.WaitGroup
+	errs := make(chan error, len(nodes)-1)
+	for _, node := range nodes {
+		if m.Node.ID == node.ID {
+			continue // Don't execute statement on ourselves.
+		}
+
+		wg.Add(1)
+		go func(node meta.NodeInfo) {
+			defer wg.Done()
+			if err := m.executeOnNode(stmt, database, &node); err != nil {
+				errs <- err
```

Comment:
Can you add information about which node failed? I didn't see any info attached in `executeOnNode()`.